rasterio/io.py

```diff
@@ -195,6 +202,66 @@ def __repr__(self):
             self.closed and 'closed' or 'open', self.name, self.mode)
 
 
+class MemoryFile(MemoryFileBase):
+    """A BytesIO-like object, backed by an in-memory file.
+
+    This allows formatted files to be read and written without I/O.
+
+    A MemoryFile created with initial bytes becomes immutable. A
+    MemoryFile created without initial bytes may be written to using
+    either file-like or dataset interfaces.
+
+    Examples
+    --------
+
+    A GeoTIFF can be loaded in memory and accessed using the GeoTIFF
+    format driver
+
+    >>> with open('tests/data/RGB.byte.tif', 'rb') as f, \
+    ...         MemoryFile(f.read()) as memfile:
+    ...     with memfile.open() as src:
+    ...         pprint.pprint(src.profile)
+    ...
+    {'count': 3,
+     'crs': CRS({'init': 'epsg:32618'}),
+     'driver': 'GTiff',
+     'dtype': 'uint8',
+     'height': 718,
+     'interleave': 'pixel',
+     'nodata': 0.0,
+     'tiled': False,
+     'transform': Affine(300.0379266750948, 0.0, 101985.0,
+           0.0, -300.041782729805, 2826915.0),
+     'width': 791}
+
+    """
+
+    def open(self, driver=None, width=None, height=None,
+             count=None, crs=None, transform=None, dtype=None, nodata=None,
+             **kwargs):
+        """Open the file and return a Rasterio dataset object."""
```

Comment:
Well, `rasterio.open(memfile.name)` works right now, but it's more verbose than `memfile.open()` and 1) lacks the proper emphasis on the `MemoryFile` as the context, and 2) surfaces the "/vsimem" name unnecessarily. I'm also using http://py.readthedocs.io/en/latest/path.html#py._path.local.LocalPath.open as a guiding light. Its `open()` method returns a Python file object and I find this to be a good analogy for Rasterio.